src/pages/Environments.js

```diff
@@ -116,8 +120,165 @@ const DeleteAppModal = ({ app: { googleProject, appName, diskName, appType }, on
+    backgroundColor: colors.danger(0.15), color: colors.danger()
+  }
+}, [
+  h(TooltipTrigger, { content: `This disk is shared between workspaces, which is no longer supported. Click "Migrate" to make copies for relevant workspaces.` }, [
+    div(['Offline', icon('warning-standard', { style: { marginLeft: '0.25rem', color: colors.danger() } })])
+  ]),
+  h(Link, { onClick: () => setMigrateDisk(disk), style: { wordBreak: 'break-word' } },
+    ['Migrate']
+  )
+])
+
+const MigratePersistentDiskModal = ({ disk, workspaces, onSuccess, onDismiss, contactSupport, deleteDiskId }) => {
+  const [migrating, setMigrating] = useState()
+  const [isSelected, setIsSelected] = useState({})
+  const [deleteDisk, setDeleteDisk] = useState(false)
+
+  const copyDiskToWorkspace =
+    ({ googleProject, namespace: saturnWorkspaceNamespace, name: saturnWorkspaceName }) => {
+      return reportErrorAndRethrow(`Error copying persistent disk to workspace '${saturnWorkspaceName}'`,
+        () => Ajax().Disks.disk(googleProject, disk.name).create({
+          ..._.pick(['size', 'blockSize', 'zone'], disk),
+          diskType: disk.diskType.label,
+          labels: { saturnWorkspaceNamespace, saturnWorkspaceName },
+          sourceDisk: _.pick(['googleProject', 'name'], disk)
+        })
+      )()
+    }
+
+  const migrateDisk = _.flow(onError(onDismiss), Utils.withBusyState(setMigrating))(async () => {
+    const destinations = _.filter(({ workspaceId }) => isSelected[workspaceId], workspaces)
+    await Promise.all(_.map(copyDiskToWorkspace, destinations))
+    await onSuccess()
+  })
+
+  const DeleteDiskSelection = ({ ...props }) => {
+    const onChange = choice => {
+      setDeleteDisk(choice)
+      if (choice) { setIsSelected({}) }
+    }
+
+    return div({ style: { ...props?.style, display: 'flex', flexDirection: 'row', alignItems: 'center' } }, [
+      h(LabeledCheckbox, { checked: deleteDisk, onChange }, []),
+      ['Delete this disk.']
+    ])
+  }
+
+  const costPerCopy = getPersistentDiskCostMonthly(disk, getRegionFromZone(disk.zone))
+  const numberOfCopies = _.flow(_.values, _.compact, _.size)(isSelected)
+
+  const space = { style: { marginTop: '0.5rem', marginBottom: '0.5rem' } }
+
+  return h(Modal, {
+    title: `Migrate ${disk.name}`,
+    okButton: deleteDisk || _.isEmpty(workspaces) ?
+      h(ButtonPrimary, { disabled: !deleteDisk, onClick: _.flow(onDismiss, _.constant(disk.id), deleteDiskId) }, 'Delete') :
+      h(ButtonPrimary, { disabled: numberOfCopies === 0, onClick: migrateDisk }, 'Migrate'),
+    onDismiss
+  }, [
+    div({ style: { display: 'flex', flexDirection: 'column' } }, Array.prototype.concat(
+      [span(space, 'Due to data security policies, persistent disks can no longer be shared between workspaces.')],
+      _.isEmpty(workspaces) ? [
+        strong(space, ['You own this disk but do not have access to any workspaces where it can be shared.']),
+        h(DeleteDiskSelection, space),
```

Comment:
Because `DeleteDiskSelection` is defined inside `MigratePersistentDiskModal`, a new function will be created on each render of `MigratePersistentDiskModal`. Thus, when it's rendered as a component like this, React will see it as a different component and unmount and destroy the previous instance and create and mount a new instance. This is inefficient and can lead to unexpected behavior if any hooks were used in `DeleteDiskSelection`

Instead of using `DeleteDiskSelection` as a component, you can use it as a function (you may want to rename it in this case).

```suggestion
        DeleteDiskSelection(space),
```

Reply:
I'd like to understand this more, in particular when to `h` stuff and when to not (you can tell I'm *clearly* very knowledgeable...). Maybe a topic to discuss in our meeting